src/target/riscv/riscv.c

```diff
@@ -5084,189 +5086,140 @@ int riscv_enumerate_triggers(struct target *target)
-			return "tdata2";
-		case GDB_REGNO_MISA:
-			return "misa";
-		case GDB_REGNO_DPC:
-			return "dpc";
-		case GDB_REGNO_DCSR:
-			return "dcsr";
-		case GDB_REGNO_DSCRATCH0:
-			return "dscratch0";
-		case GDB_REGNO_MSTATUS:
-			return "mstatus";
-		case GDB_REGNO_MEPC:
-			return "mepc";
-		case GDB_REGNO_MCAUSE:
-			return "mcause";
-		case GDB_REGNO_PRIV:
-			return "priv";
-		case GDB_REGNO_SATP:
-			return "satp";
-		case GDB_REGNO_VTYPE:
-			return "vtype";
-		case GDB_REGNO_VL:
-			return "vl";
-		case GDB_REGNO_V0:
-			return "v0";
-		case GDB_REGNO_V1:
-			return "v1";
-		case GDB_REGNO_V2:
-			return "v2";
-		case GDB_REGNO_V3:
-			return "v3";
-		case GDB_REGNO_V4:
-			return "v4";
-		case GDB_REGNO_V5:
-			return "v5";
-		case GDB_REGNO_V6:
-			return "v6";
-		case GDB_REGNO_V7:
-			return "v7";
-		case GDB_REGNO_V8:
-			return "v8";
-		case GDB_REGNO_V9:
-			return "v9";
-		case GDB_REGNO_V10:
-			return "v10";
-		case GDB_REGNO_V11:
-			return "v11";
-		case GDB_REGNO_V12:
-			return "v12";
-		case GDB_REGNO_V13:
-			return "v13";
-		case GDB_REGNO_V14:
-			return "v14";
-		case GDB_REGNO_V15:
-			return "v15";
-		case GDB_REGNO_V16:
-			return "v16";
-		case GDB_REGNO_V17:
-			return "v17";
-		case GDB_REGNO_V18:
-			return "v18";
-		case GDB_REGNO_V19:
-			return "v19";
-		case GDB_REGNO_V20:
-			return "v20";
-		case GDB_REGNO_V21:
-			return "v21";
-		case GDB_REGNO_V22:
-			return "v22";
-		case GDB_REGNO_V23:
-			return "v23";
-		case GDB_REGNO_V24:
-			return "v24";
-		case GDB_REGNO_V25:
-			return "v25";
-		case GDB_REGNO_V26:
-			return "v26";
-		case GDB_REGNO_V27:
-			return "v27";
-		case GDB_REGNO_V28:
-			return "v28";
-		case GDB_REGNO_V29:
-			return "v29";
-		case GDB_REGNO_V30:
-			return "v30";
-		case GDB_REGNO_V31:
-			return "v31";
-		default:
-			if (regno <= GDB_REGNO_XPR31)
-				sprintf(buf, "x%d", regno - GDB_REGNO_ZERO);
-			else if (regno >= GDB_REGNO_CSR0 && regno <= GDB_REGNO_CSR4095)
-				sprintf(buf, "csr%d", regno - GDB_REGNO_CSR0);
-			else if (regno >= GDB_REGNO_FPR0 && regno <= GDB_REGNO_FPR31)
-				sprintf(buf, "f%d", regno - GDB_REGNO_FPR0);
+	int read_symbols = snprintf(buf->name, size_buf, "%s", name_prefix);
+	assert(read_symbols > 0 && "register name is too long");
+}
+
+static void initialize_reg_name_with_prefixed_name(struct reg_name *buf, const char *name_prefix,
+	unsigned int num_in_name)
+{
+	const unsigned int size_buf = ARRAY_SIZE(buf->name);
+
+	int read_symbols = snprintf(buf->name, size_buf, "%s%d", name_prefix, num_in_name);
+	assert(read_symbols > 0 && "register name is too long");
+}
+
+const char *gdb_regno_name(struct target *target, enum gdb_regno regno)
+{
+	if (regno >= GDB_REGNO_COUNT) {
+		RISCV_INFO(info);
+		assert(info->custom_register_names.reg_names);
+		assert(regno - GDB_REGNO_COUNT <= info->custom_register_names.names_num);
+		return info->custom_register_names.reg_names[regno - GDB_REGNO_COUNT].name;
+	}
+
+	static struct reg_name buf[GDB_REGNO_COUNT] = {
```

Comment:
Please, do not use static local variables (whose lifetime and point of initialization does not match their scope).

Instead, please move this buffer to `struct riscv_info`. 

Initialize it in `riscv_init_target()` and de-initialize in `riscv_deinit_target()`. 

This way the point of creation of the data becomes well defined (at target initialization, as opposed to first call to gdb_regno_name()). Similarly, the validity period of the data will become clear (from target initialization till the de-initialization).

Thanks.

Reply:
In this case, local variables are used so as not to allocate memory for each target, since the data is the same.

Reply:
static makes sense to me. Really this is data that could be hard-coded in the source, but instead it's computed once during program execution.

Reply:
> AnastasiyaChernikova: [...] for each target, since the data is the same.

> timsifive: [...] Really this is data that could be hard-coded in the source, but instead it's computed once during program execution.

Thank you for explanation. 

In that case, please use a global const array, initialized with the constant data. Thearray can even go to a separate `.h / .c` files if you wish (e.g. `riscv_reg_names.h`, `riscv_reg_names.c`). 

That way:

- the register names will be available to any user, if needed (not just `gdb_regno_name()`);
- the register names will be constant (immutable);
- it will be more clear to the reader that these are fixed constants (known at compile-time), not something that needs to be dynamically computed;
- the somewhat weird pattern of using static variables inside functions (whose lifetime and initialization does not correspond to their scope, as already noted above) is avoided

Reply:
I would like to avoid creating such a file, since it will be hard to maintain. Please, take a look at the new approach.

Reply:
Marking a resolved - outdated thread.